src/main/java/com/github/dbmdz/solrocr/reader/BaseSourceReader.java

```diff
@@ -0,0 +1,242 @@
+    }
+    if (start + byteLen > this.length()) {
+      byteLen = this.length() - start;
+    }
+    byte[] data = new byte[byteLen];
+    this.readBytes(data, 0, start, byteLen);
+    int dataStart = adjustOffset(0, data, AdjustDirection.RIGHT);
+    int dataEnd = adjustOffset(data.length - 1, data, AdjustDirection.LEFT);
+    return new String(data, dataStart, dataEnd - dataStart + 1, StandardCharsets.UTF_8);
+  }
+
+  /**
+   * Move offset if we're on a partital UTF8 multi-byte sequence so that we start at a valid UTF8
+   * sequence
+   */
+  private static int adjustOffset(int offset, byte[] buf, AdjustDirection direction) {
+    int b = buf[offset] & 0xFF;
+    if (direction == AdjustDirection.RIGHT) {
+      while ((b >> 6) == 0b10) {
+        offset += 1;
+        if (offset >= buf.length) {
+          return -1;
+        }
+        b = buf[offset] & 0xFF;
+      }
+    } else {
+      int hiBits = (b >> 6);
+      boolean isMultiByte = hiBits == 0b11 || hiBits == 0b10;
+      if (!isMultiByte) {
+        return offset;
+      }
+      int shift = 0;
+      while (offset - shift >= 0) {
+        shift++;
+        b = buf[offset - shift] & 0xFF;
+        hiBits = b >> 6;
+        if (hiBits == 0b11) {
+          if (b >> 5 == 0b110 && shift == 1) {
+            // Complete 2-byte sequence
+            return offset;
+          } else if (b >> 4 == 0b1110 && shift == 2) {
+            // Complete 3-byte sequence
+            return offset;
+          } else if (b >> 3 == 0b11110 && shift == 3) {
+            // Complete 4-byte sequence
+            return offset;
+          } else {
+            // Unfinished sequence
+            return offset - shift - 1;
+          }
+        } else if (hiBits < 2) {
+          // Non-multibyte character, adjust to herer
+          return offset - shift;
+        }
+      }
+    }
+    return offset;
+  }
+
+  public Section getAsciiSection(int offset) throws IOException {
+    if (offset < 0) {
+      throw new IllegalArgumentException("offset must be >= 0");
+    }
+    if (offset >= this.length()) {
+      throw new IllegalArgumentException("offset must be < length");
+    }
+    int sectionIndex = offset / sectionSize;
+    if (cache == null) {
+      initializeCache();
+    }
+    if (cache[sectionIndex] != null) {
+      cache[sectionIndex].lastUsedTimestampNs = System.nanoTime();
+      return cache[sectionIndex].section;
+    }
+    int startOffset = sectionIndex * sectionSize;
+    int readLen = Math.min(sectionSize, this.length() - startOffset);
+    this.readBytes(copyBuf, 0, startOffset, readLen);
+    // Construct a String without going through a decoder to save on CPU.
+    // Given that the method has been deprecated since Java 1.1 and was never removed, I don't think
+    // this is very risky 😅
+    Section section =
+        new Section(startOffset, startOffset + sectionSize, new String(copyBuf, 0, 0, readLen));
+    if (cache.length > 0 && cacheSlotsUsed == cache.length) {
```

Comment:
I think we need to check against `maxCacheEntries`:
```suggestion
    if (cache.length > 0 && cacheSlotsUsed == maxCacheEntries) {
```